services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/KafkaBasedCommandConsumerFactoryImpl.java

```diff
@@ -0,0 +1,240 @@
+        // The consumer is created already in start method as it uses topic pattern (irrespective of tenant-id) to
+        // subscribe.
+        return Future.succeededFuture();
+    }
+
+    @Override
+    public Future<Void> start() {
+        final Promise<Void> subscriptionPromise = Promise.promise();
+        final Map<String, String> consumerConfig = kafkaConsumerConfig.getConsumerConfig();
+
+        consumerConfig.computeIfAbsent(ConsumerConfig.GROUP_ID_CONFIG, k -> componentName);
+        kafkaConsumerFactory.getOrCreateConsumer(componentName, consumerConfig)
+                .subscribe(topicPatternForIncomingCommands, subscriptionPromise)
+                .handler(this::mapAndDelegateIncomingCommandMessage);
+
+        return subscriptionPromise.future()
+                .map(ok -> {
+                    log.info("Kafka consumer subscribed to the topic [{}]", topicPatternForIncomingCommands);
+                    return null;
+                });
+    }
+
+    @Override
+    public Future<Void> stop() {
+        return CompositeFuture.all(super.stop(), kafkaConsumerFactory.closeConsumer(componentName))
+                .mapEmpty();
+    }
+
+    private void mapAndDelegateIncomingCommandMessage(final KafkaConsumerRecord<String, Buffer> incomingCommandRecord) {
+        Objects.requireNonNull(incomingCommandRecord);
+
+        final Command command = Command.from(incomingCommandRecord);
+        final Span currentSpan = tracer.buildSpan("map and delegate command")
+                .ignoreActiveSpan()
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+                .withTag(TracingHelper.TAG_TENANT_ID, command.getTenantId())
+                .withTag(TracingHelper.TAG_DEVICE_ID, command.getDeviceId())
+                .start();
+
+        log.debug("map and delegate command [commandTargetMapper: {}]", commandTargetMapper);
+        if (!command.isValid()) {
+            log.debug("command is invalid");
+            TracingHelper.logError(currentSpan, "command is invalid");
+            return;
+        }
+
+        commandTargetMapper
+                .getTargetGatewayAndAdapterInstance(command.getTenantId(), command.getDeviceId(), currentSpan.context())
+                .onSuccess(result -> {
+                    final String targetAdapterInstanceId = result
+                            .getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+                    final String targetDeviceId = result.getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                    final String gatewayId = targetDeviceId.equals(command.getDeviceId()) ? null : targetDeviceId;
+                    forwardCommand(command, gatewayId, targetAdapterInstanceId, incomingCommandRecord.value(),
+                            currentSpan);
+                    // TODO to explore options for committing manually based on the QoS once the command is
+                    // successfully forwarded
```

Comment:
You cannot mix AT MOST ONCE and AT LEAST ONCE semantics in one consumer. The consumer either commits the offset *before* (AT MOST ONCE) or *after* (AT LEAST ONCE) processing the message. To be able to decide this based on a property of the message, we would have to fetch each message individually from the broker (opposed to fetching batches) which is not desirable. In other words, you have to decide if messages should be consumed twice or if you accept message loss if the consumer is interrupted or an error occurs. We decided to live with duplicates because the application layer must be able to handle them anyway. Long story short, I would always commit the offsets after processing the message. This is what `AbstractAtLeastOneKafkaConsumer` already provides (see #2420).